text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+alternative on the basis that they serve similar functions, rather than
+proposing that they can and should coexist. This RFC proposes that we allow
+`let`-chaining *and* add the `is` operator.
+
+`if`-`let` provides a natural extension of `let` for fallible bindings, and
+highlights the binding by putting it on the left, like a `let` statement.
+Allowing developers to chain multiple `let` operations and other expressions in
+the `if` condition provides a natural extension that simplifies what would
+otherwise require complex nested conditionals. As the `let`-chains RFC notes,
+this is a feature people already expect to work.
+
+The `is` operator similarly allows developers to chain multiple match-and-bind
+operations and simplify what would otherwise require complex nested
+conditionals. However, the `is` operator allows writing and reading a pattern
+match from left-to-right, which reads more naturally in many circumstances. For
+instance, consider an expression like `x is Some(y) && y > 5`; that boolean
+expression reads more naturally from left-to-right than
+`let Some(y) = x && y > 5`.
+
+This is even more true at the end of a longer expression chain, such as
+`x.method()?.another_method().await? is Some(y)`. Rust method chaining and `?`
+and `.await` all encourage writing code that reads in operation order from left
+to right, and `is` fits naturally at the end of such a sequence.
+
+Having an `is` operator would also help to reduce the proliferation of methods
+on types such as `Option` and `Result`, by allowing prospective users of those
+methods to write a condition using `is` instead. While any such condition could
+equivalently be expressed using `let`-chains, the binding would then move
+further away from the condition expression referencing the binding, which would
+result in a less natural reading order for the expression.
+
+Consider the following examples:
+
+```rust
+if expr_producing_option().is_some_and(|v| condition(v))
+
+if let Some(v) = expr_producing_option() && condition(v)
+
+if expr_producing_option() is Some(v) && condition(v)
+```
+
+The condition using `is` is a natural translation from the `is_some_and`
+method, whereas the if-let construction requires reversing the binding of `v`
+and the expression producing the option. This seems sufficiently cumbersome in
+some cases that the absence of `is` would motivate continued use and
+development of helper methods.
+
+# Guide-level explanation
+
+Rust provides an `is` operator, which can be used in any expression:
+`EXPR is PATTERN`
+
+This operator tests if the value of `EXPR` matches the specified `PATTERN`; see
+<https://doc.rust-lang.org/reference/patterns.html> for details on patterns.
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
```

Comment:
> known

In all examples, it's obvious and I'm sure the compiler can figure it out, but I can trivially write code examples where it's non-obvious whether the value is true

```rust
let is_true = x is Some(y);
let is_true = identity(is_true);
if is_true { y; }
```
What happens here? What about just having a local variable? The compiler needs to draw a line somewhere (as Rust is turing complete), and this doesn't say anything about what that is. I think no matter where it lies, it can be pretty confusing to users.

Reply:
I get the impression from the RFC text that _probably_ it's not intended for the binding `y` to live beyond the end of the first statement, as the second statement is potentially reachable when `x is None`.

Reply:
Yeah, I gathered that too and didn't realise that it wasn't stated explicitly. To put it more clearly, `x is y` only binds `y` for the current expression in the tree, *at the same level*, meaning that `(x is y) == ...` will not allow `y` to occur inside `...`.

Reply:
While I do not have a solution it does feel like the binding should last longer than just the expression it is within. Else this Rust equivalent of C# would not be possible:
```csharp
if(foo is {Values: [var value1, var value2]}) {
    return (value1, value2);
}
```